File: tools/util/minifind.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fnmatch.h>
#include <errno.h>

enum FileType { FT_ANY = 0, FT_FILE, FT_DIR };

struct Options {
    enum FileType want_type;
    const char *name_pattern;
};

static void usage(const char *prog) {
    fprintf(stderr, "Usage: %s <path> [-type f|d] [-name pattern]\n", prog);
    exit(1);
}

static int match_type(const struct stat *st, enum FileType t) {
    if (t == FT_ANY) return 1;
    if (t == FT_FILE && S_ISREG(st->st_mode)) return 1;
    if (t == FT_DIR && S_ISDIR(st->st_mode)) return 1;
    return 0;
}

static void traverse(const char *path, struct Options *opt) {
    struct stat st;
    if (lstat(path, &st) < 0) {
        perror(path);
        return;
    }

    const char *basename = strrchr(path, '/');
    basename = basename ? (basename + 1) : path;

    int name_ok = 1;
    if (opt->name_pattern) {
        if (fnmatch(opt->name_pattern, basename, 0) != 0) {
            name_ok = 0;
        }
    }

    if (name_ok && match_type(&st, opt->want_type)) {
        printf("%s\n", path);
    }

    if (S_ISDIR(st.st_mode)) {
        DIR *dp = opendir(path);
        if (!dp) {
            perror(path);
            return;
        }
        struct dirent *de;
        while ((de = readdir(dp))) {
            if (strcmp(de->d_name, ".") == 0 || strcmp(de->d_name, "..") == 0)
                continue;
            char buf[PATH_MAX];
            snprintf(buf, sizeof(buf), "%s/%s", path, de->d_name);
            traverse(buf, opt);
        }
        closedir(dp);
    }
}

int main(int argc, char *argv[]) {
    if (argc < 2) usage(argv[0]);

    struct Options opt = { FT_ANY, NULL };
    const char *start = NULL;

    // simple parsing: first non-option is path
    int i;
    for (i = 1; i < argc; i++) {
        if (argv[i][0] != '-') {
            if (!start) start = argv[i];
            else usage(argv[0]);
        } else if (strcmp(argv[i], "-type") == 0) {
            if (i + 1 >= argc) usage(argv[0]);
            i++;
            if (argv[i][0] == 'f') opt.want_type = FT_FILE;
            else if (argv[i][0] == 'd') opt.want_type = FT_DIR;
            else usage(argv[0]);
        } else if (strcmp(argv[i], "-name") == 0) {
            if (i + 1 >= argc) usage(argv[0]);
            i++;
            opt.name_pattern = argv[i];
        } else {
            usage(argv[0]);
        }
    }

    if (!start) usage(argv[0]);
    traverse(start, &opt);
    return 0;
}

File: tools/util/minifind.py
import os
import fnmatch
import sys

class FileType:
    ANY = 0
    FILE = 1
    DIR = 2

class Options:
    def __init__(self):
        self.want_type = FileType.ANY
        self.name_pattern = None

def usage(prog):
    print(f"Usage: {prog} <path> [-type f|d] [-name pattern]")
    sys.exit(1)

def match_type(path, file_type):
    """Match the file type (regular file or directory)."""
    if file_type == FileType.ANY:
        return True
    try:
        stat_info = os.lstat(path)
        if file_type == FileType.FILE and os.path.isfile(path):
            return True
        if file_type == FileType.DIR and os.path.isdir(path):
            return True
    except OSError:
        return False
    return False

def traverse(path, opt):
    """Traverse the directory and print matching files."""
    try:
        stat_info = os.lstat(path)
        basename = os.path.basename(path)
        name_ok = True

        # Check if the basename matches the pattern
        if opt.name_pattern and fnmatch.fnmatch(basename, opt.name_pattern) == False:
            name_ok = False

        # If the name matches and the type is correct, print the path
        if name_ok and match_type(path, opt.want_type):
            print(path)

        # If it's a directory, recursively traverse it
        if os.path.isdir(path):
            try:
                for entry in os.listdir(path):
                    full_path = os.path.join(path, entry)
                    if entry != '.' and entry != '..':
                        traverse(full_path, opt)
            except OSError:
                pass

    except OSError as e:
        print(f"Error: {e.strerror} - {path}")

def main():
    if len(sys.argv) < 2:
        usage(sys.argv[0])

    # Initialize the options
    opt = Options()
    start = None

    # Parse arguments
    i = 1
    while i < len(sys.argv):
        if sys.argv[i][0] != '-':
            if not start:
                start = sys.argv[i]
            else:
                usage(sys.argv[0])
        elif sys.argv[i] == "-type":
            if i + 1 >= len(sys.argv):
                usage(sys.argv[0])
            i += 1
            if sys.argv[i] == "f":
                opt.want_type = FileType.FILE
            elif sys.argv[i] == "d":
                opt.want_type = FileType.DIR
            else:
                usage(sys.argv[0])
        elif sys.argv[i] == "-name":
            if i + 1 >= len(sys.argv):
                usage(sys.argv[0])
            i += 1
            opt.name_pattern = sys.argv[i]
        else:
            usage(sys.argv[0])
        i += 1

    if not start:
        usage(sys.argv[0])

    traverse(start, opt)

if __name__ == "__main__":
    main()

